PZNS_Framework/media/lua/client/03_mod_core/PZNS_NPCGroup.lua

```diff
@@ -1,22 +1,93 @@
-PZNS_NPCGroup = {};
-PZNS_NPCGroup.__index = PZNS_NPCGroup;
-
---- Cows: can set the group leader later.
----@param groupID any
----@return table
-function PZNS_NPCGroup:newGroup(
-    groupID
+require("00_references/init")
+
+---@class Group
+---@field groupID groupID           Unique group identifier
+---@field name string               Group name
+---@field leaderID survivorID       NPC ID of group leader
+---@field members table<survivorID> list of group members
+---@field memberCount integer       number of members in group
+---@field factionID factionID?      Faction this group belongs to
+local Group = {}
+
+-- - Cows: can set the group leader later.
+--- Creates new group with provided parameters and puts it to registry
+---@param groupID groupID
+---@param name string
+---@param leaderID survivorID
+---@param members? table<survivorID>
+---@param factionID? factionID
+---@return Group
+function Group:new(
+    groupID,
+    name,
+    leaderID,
+    members,
+    factionID
 )
-    local npcGroup = {};
-    setmetatable(npcGroup, self)
+    local group = {};
+    group.groupID = groupID
+    group.name = name
+    group.leaderID = leaderID
+    group.members = members or {}
+    group.factionID = factionID
+    group.memberCount = #group.members
+    setmetatable(group, self)
     self.__index = self;
+    return group;
+end
 
-    npcGroup = {
-        groupID = groupID,
-        groupLeader = "",  -- Cows: Probably useless... since only the player can actually "lead" by giving commands and designate zones.
-        -- groupMembers = "", -- Cows: Probably useless... since we can map npcIDs as keys and values to the group.
-        -- groupHomeZone = "" --  Cows: Probably useless... Zones are tied to the groupID.
-    };
+---Sets new leader for group
+---@param newLeaderID survivorID survivorID of new leader
+function Group:setLeader(newLeaderID)
+    self.leaderID = newLeaderID
+end
 
-    return npcGroup;
+---Get list of members (survivorID)
+---@return table<survivorID>
+function Group:getMembers()
+    local res = {}
+    for i = 1, #self.members do
+        res[i] = self.members[i]
+    end
+    return res
 end
+
+---Get number of group members
+function Group:getMemberCount()
+    return self.memberCount
+end
+
+---Added memberID to group members
+---@param memberID survivorID NPC survivorID of newly added member
+function Group:addMember(memberID)
+    if not memberID then return end
+    self.members[self.memberCount + 1] = memberID
+    self.memberCount = self.memberCount + 1
+end
+
+---Removes <memberID> from Group members
+---@param memberID survivorID NPC survivorID of member to remove
+function Group:removeMember(memberID)
+    if not memberID then return end
+    if memberID == self.leaderID then
+        self.leaderID = nil
+    end
+    for i = 1, #self.members do
+        if self.members[i] == memberID then
+            table.remove(self.members, i)
+            self.memberCount = self.memberCount - 1
```

Comment:
Same question, see above in Faction.

If memberCount changes in-game due to NPCs removed from group, what happens to the list index?